client/apps/edit/components/content/sections/text/index.jsx

```diff
@@ -0,0 +1,557 @@
+  }
+
+  checkSelection = () => {
+    const { hasFeatures } = this.props
+    const { showUrlInput } = this.state
+
+    if (this.domEditor && !showUrlInput) {
+      if (this.hasSelection()) {
+        const editor = ReactDOM.findDOMNode(this.domEditor)
+        const editorPosition = editor.getBoundingClientRect()
+        const selectionLeft = hasFeatures ? 125 : 100
+        const selectionTarget = stickyControlsBox(editorPosition, -93, selectionLeft)
+
+        this.setState({
+          showMenu: true,
+          selectionTarget
+        })
+      } else {
+        this.setState({showMenu: false})
+      }
+    }
+  }
+
+  render () {
+    const {
+      article,
+      editing,
+      hasFeatures,
+      isContentStart
+    } = this.props
+    const {
+      editorState,
+      selectionTarget,
+      showMenu,
+      showUrlInput,
+      plugin,
+      url
+    } = this.state
+
+    const blocks = Config.blockTypes(article.layout, hasFeatures)
+    const blockRenderMap = Config.blockRenderMap(article.layout, hasFeatures)
+    const styles = Config.inlineStyles(article.layout, hasFeatures)
+    const showDropCaps = editing ? false : isContentStart
+
+    return (
+    <div
+      className='SectionText'
+      data-editing={editing}
+      onClick={this.focus}
+    >
+      <Text
+        layout={article.layout}
+        isContentStart={showDropCaps}
+      >
+        {showMenu &&
+          <TextNav
+            blocks={blocks}
+            hasFeatures={hasFeatures}
+            makePlainText={this.makePlainText}
+            promptForLink={this.promptForLink}
+            styles={styles}
+            toggleBlock={this.toggleBlock}
+            toggleStyle={this.toggleStyle}
+            position={selectionTarget}
+          />
+        }
+        <div
+          className='SectionText__input'
+          onMouseUp={this.checkSelection}
+          onKeyUp={this.checkSelection}
+        >
+          <Editor
+            ref={(ref) => { this.domEditor = ref }}
+            blockRenderMap={blockRenderMap}
+            decorators={Config.decorators}
+            editorState={editorState}
+            handleKeyCommand={this.handleKeyCommand}
+            handlePastedText={this.onPaste}
+            handleReturn={this.handleReturn}
+            keyBindingFn={keyBindingFnFull}
+            onChange={this.onChange}
+            onTab={this.handleTab}
+            onLeftArrow={this.handleChangeSection}
+            onUpArrow={this.handleChangeSection}
+            onRightArrow={this.handleChangeSection}
+            onDownArrow={this.handleChangeSection}
+            spellcheck
+          />
+        </div>
+
+        {editing && showUrlInput &&
+          <TextInputUrl
+            removeLink={this.removeLink}
+            confirmLink={this.confirmLink}
+            selectionTarget={selectionTarget}
+            pluginType={plugin}
+            urlValue={url}
+          />
+        }
+
+      </Text>
+    </div>
+    )
+  }
+}
```

Comment:
My general feedback on this component is that this is a _lot_ of logic. Generally when a file is getting to be over 500 lines, I would consider breaking it down for easier reading, testing, and consumption. I see a few opportunities to extract some of this into sub components. Thinking something along the lines of `<TextNavWrapper/>` and `<EditorWrapper/>`. Thoughts?

Reply:
It grew by ~200 lines when I converted it to JSX, but actually now has less logic than it did before-- moved a number of other functions into the utils files, which could probably go further.  I don't think this can break into any more subcomponents ATM.

Reply:
Could you explain why it couldn't be broken into more subcomponents?

Happy that it has less logic than it did before, but looking holistically, it's still a lot for one file. Most of those methods get passed into components through props and could probably work with an subcomponent that was responsible for managing changes between the two. I'm not saying that it has to be done now or even at all, but instead giving some feedback on the readability of this file from a reviewer's perspective.

Reply:
It can't be broken into components because there are no dom elements left that aren't already their own component except the wrapper, but that needs access to the editor ref to focus the editor correctly -- they all have to live side by side except reaction's Text component.  Working on moving more things into Utils, which puts it closer to 400 lines.

Reply:
Also we can potentially connect the subcomponents once the new redux setup is up and running, which would remove some of the need to pass props.

Reply:
Not sure I understand. Wrapper components don't need to have additional dom elements -- we can use them to extract methods and pass the original component into the render method:

```
export class TextInputUrlWrapper extends Component {

   componentWillUpdate() {
      // maybe do some operations to keep state/parents in sync
   }

   removeLink: () => {
      const newEditorState = // get new editorState after removing link
      this.props.onEditorStateChange(newEditorState) //tell parent about change
   }

   render () {
       const {
         selectionTarget,
         plugin,
         url
       } = this.props

       return (
         <TextInputUrl
           removeLink={this.removeLink}
           confirmLink={this.confirmLink}
           selectionTarget={selectionTarget}
           pluginType={plugin}
           urlValue={url}
         />
       )
   }
}
```

I might still be missing something but keeping data like `editorState` and references to the editor in sync across different components should be possible (and likely much easier with Redux). Again, this is more of a response to the readability/responsibility of the component and it's not a blocker in any way but I'd be happy to continue the discussion and pair on this at some point!